src/soter/openssl/soter_asym_ka.c

```diff
@@ -170,25 +165,37 @@ soter_status_t soter_asym_ka_import_key(soter_asym_ka_t* asym_ka_ctx, const void
         return SOTER_INVALID_PARAMETER;
     }
 
-    pkey = EVP_PKEY_CTX_get0_pkey(asym_ka_ctx->pkey_ctx);
-
-    if (!pkey) {
-        return SOTER_INVALID_PARAMETER;
-    }
-
-    if (EVP_PKEY_EC != EVP_PKEY_id(pkey)) {
-        return SOTER_INVALID_PARAMETER;
+    /*
+     * soter_ec_{priv,pub}_key_to_engine_specific() expect EVP_PKEY of EVP_PKEY_EC type
+     * to be already allocated and non-NULL. We might be importing it anew, or we might be
+     * replacing previously generated keypair.
+     */
+    if (asym_ka_ctx->pkey) {
+        if (EVP_PKEY_base_id(asym_ka_ctx->pkey) != EVP_PKEY_EC) {
+            return SOTER_INVALID_PARAMETER;
+        }
+    } else {
+        asym_ka_ctx->pkey = EVP_PKEY_new();
```

Comment:
why we don't use here `soter_asym_ka_gen_key` function to generate new empty key?

Reply:
> why we don't use here soter_asym_ka_gen_key function to generate new empty key?

It does not generate an _empty_ key, it generates a new keypair. `soter_asym_ka_import_key()` imports either a private or public key into existing keypair. Generating a new keypair right before importing a half of it does not really make sense. So we only create a blank EVP_PKEY to import into, just because `soter_ec_{priv,pub}_key_to_engine_specific()` don't do it and expect EVP_PKEY to be already allocated.